src/execution/linear_memory.rs

```diff
@@ -446,68 +446,66 @@ impl<const PAGE_SIZE: usize> LinearMemory<PAGE_SIZE> {
 
 impl<const PAGE_SIZE: usize> core::fmt::Debug for LinearMemory<PAGE_SIZE> {
     fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
-        write!(f, "LinearMemory {{ inner_data: [ ")?;
-        let lock_guard = self.inner_data.read();
-
-        // then write all other elements with a preceding ", "
-        let mut repetition_count = 1;
-
-        // if this many or more sucessive elements have the same value, print just the number of
-        // repetitions and one time the element value
-        let repetition_threshold = 8;
-
-        for i in 1..lock_guard.len() {
-            // SAFETY: As `i` is a valid index into this `LinearMemory` while the
-            // `lock_guard` prevents a resize/realloc of its data, so its valid.
-            let prev_byte = unsafe { *lock_guard[i - 1].get() };
-
-            // SAFETY: As `i + 1` is a valid index into this `LinearMemory` while the
-            // `lock_guard` prevents a resize/realloc of its data, so its valid.
-            let byte = unsafe { *lock_guard[i].get() };
-
-            let is_last_element = i == lock_guard.len() - 1;
-            let is_repetition = prev_byte == byte;
-
-            // if this is a repetition, increment the repetition_count
-            if is_repetition {
-                repetition_count += 1;
-            }
-
-            // either the repition was broken or we are at the last element, so write the repitition
-            if !is_repetition || is_last_element {
-                // threshold crossed, just print the count and the element once
-                if repetition_count >= repetition_threshold {
-                    write!(f, "#{repetition_count} × {prev_byte}")?;
-                    repetition_count = 1;
-                }
-                // threshold not crossed, actually print the repeating element count times
-                else {
-                    for _ in 0..repetition_count - 1 {
-                        write!(f, "{prev_byte}, ")?;
+        struct RepetitionDetectingMemoryWriter<'a>(ReadLockGuard<'a, Vec<UnsafeCell<u8>>>);
+        impl core::fmt::Debug for RepetitionDetectingMemoryWriter<'_> {
+            fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
+                let mut list = f.debug_list();
+
+                // allows safe access to memory
+                let read_memory = |i: usize| {
+                    assert!(i < self.0.len());
```

Comment:
This `assert` is redundant, the same assert already is in the implementation of the index operator. I would vow for removing this redundant `assert`.